aten/src/ATen/core/dispatch/KernelFunction.h

```diff
@@ -2,15 +2,208 @@
+    static_assert(guts::is_function_type<FuncType>::value, ""); // TODO
+    static_assert(!std::is_same<FuncType, BoxedKernelFunction>::value, ""); // TODO
+    static_assert(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<typename detail::WrapKernelFunction<FuncType, func>::type>()
+    );
+  }
+
+  template<class FuncType, FuncType* func>
+  static KernelFunction makeFromUnboxedOnlyFunction() {
+    // TODO We want to get rid of kernels that have only an unboxed function pointer.
+    //      All kernels should have a boxed pointer.
+
+    static_assert(guts::is_function_type<FuncType>::value, ""); // TODO
+    static_assert(!std::is_same<FuncType, BoxedKernelFunction>::value, ""); // TODO
+    static_assert(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedOnlyFunctor(
+      std::make_shared<typename detail::WrapKernelFunction<FuncType, func>::type>()
+    );
+  }
+
+  template<class FuncType, bool AllowLegacyTypes = false>
+  static KernelFunction makeFromUnboxedRuntimeFunction(FuncType* func) {
+    static_assert(guts::is_function_type<FuncType>::value, ""); // TODO
+    static_assert(!std::is_same<FuncType, BoxedKernelFunction>::value, ""); // TODO
+    TORCH_INTERNAL_ASSERT(func != nullptr, "Kernel function cannot be nullptr");
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<FuncType>>>(func)
+    );
+  }
+
+  template<class Lambda, bool AllowLegacyTypes = false>
+  static KernelFunction makeFromUnboxedLambda(Lambda&& lambda) {
+    static_assert(guts::is_functor<Lambda>::value, ""); // TODO
+    static_assert(guts::is_stateless_lambda<guts::decay_t<Lambda>>::value, ""); // TODO
+
+    return makeFromUnboxedFunctor<AllowLegacyTypes>(
+      std::make_shared<detail::WrapRuntimeKernelFunctor<guts::decay_t<Lambda>>>(std::forward<Lambda>(lambda))
+    );
+  }
+
+private:
+
+  template<class Return, class... Args>
+  Return boxAndCallBoxedFunc_(Args... args) const {
+    TORCH_INTERNAL_ASSERT(boxed_kernel_func_ != nullptr, "Tried to call KernelFunction::callUnboxed() on an uninitizliaed KernelFunction.");
+
+    // TODO Reuse stack vector instead of allocating?
+    std::vector<IValue> stack {std::forward<Args>(args)...};
+
+    (*boxed_kernel_func_)(&stack, functor_.get());
+
+    TORCH_INTERNAL_ASSERT(stack.size() == 1, "A boxed kernel should only push one return to the stack");
+    return std::move(stack[0]).to<Return>();
+  }
+
+  explicit KernelFunction(std::shared_ptr<OperatorKernel> functor, BoxedKernelFunction* boxed_kernel_func, void* unboxed_kernel_func, c10::optional<uint64_t> signature_hash)
+  : functor_(std::move(functor))
+  , boxed_kernel_func_(boxed_kernel_func)
+  , unboxed_kernel_func_(unboxed_kernel_func)
+  , signature_hash_(signature_hash)
+  {}
+
+  std::shared_ptr<OperatorKernel> functor_;
```

Comment:
What's the `functor_` here?

Reply:
That's the replacement for the old cache mechanism. It stores the actual kernel. If the kernel is registered as a functor, that functor is stored in here. If a kernel is registered as a runtime function pointer, we create a functor storing that pointer and store that functor in here.